### ISSUE TITLE
Array reference counting crash

### ISSUE THREAD

**Issue by Norbyte** [facebook/hhvm#5491]
The following code makes HHVM crash when the JIT is enabled (does not crash with -v Eval.Jit=false):
```php
class Nasty
{
    public $items;
    public function getItems()
    {
        if ($this->items === null) {
            $this->items = [];
            $this->items['X'] = 1;
        }

        return $this->items;
    }
}

function crash()
{
    $obj = new Nasty();
    $dummy = (object)['type' => 'AA'];
    $obj->getItems()[$dummy->type] = 'XXX';
    foreach ($obj->items as $account) {}
}
```

HHVM version:
HipHop VM 3.8.0-dev (dbg)
Compiler: heads/master-0-gc8420a83e032de58554d519f5ac9852e86e3025f
Repo schema: 09cfeda985fb02691d5bc57d612a3a6e9cadf30c

Error:
```hhvm/hphp/runtime/base/array-data.h:85: HPHP::RefCount HPHP::ArrayData::getCount() const: Assertion `check_refcount(m_hdr.count)' failed.```

Stack trace:
```
#0  0x00007fffee7685c9 in raise () from /lib64/libc.so.6
#1  0x00007fffee769cd8 in abort () from /lib64/libc.so.6
#2  0x00007fffee761536 in __assert_fail_base () from /lib64/libc.so.6
#3  0x00007fffee7615e2 in __assert_fail () from /lib64/libc.so.6
#4  0x00000000015f845b in getCount (this=<optimized out>) at /home/norbyte/hhvm/hhvm/hphp/runtime/base/array-data.h:85
#5  operator() (__closure=<optimized out>) at /home/norbyte/hhvm/hhvm/hphp/runtime/base/tv-helpers.cpp:59
#6  HPHP::cellIsPlausible (cell=...) at /home/norbyte/hhvm/hhvm/hphp/runtime/base/tv-helpers.cpp:77
#7  0x00000000015f8608 in HPHP::tvIsPlausible (tv=...) at /home/norbyte/hhvm/hhvm/hphp/runtime/base/tv-helpers.cpp:89
#8  0x00000000019b8f40 in tvRefcountedIncRef (tv=0x7fffd7807228) at /home/norbyte/hhvm/hhvm/hphp/runtime/base/tv-helpers.h:185
#9  HPHP::jit::MInstrHelpers::cGetPropImpl<(HPHP::KeyType)2, true> (ctx=0x0, base=<optimized out>, key=<optimized out>, mis=0x7fffcfffe030) at /home/norbyte/hhvm/hhvm/hphp/runtime/vm/jit/minstr-helpers.h:128
#10 0x00000000019b8faf in HPHP::jit::MInstrHelpers::cGetPropSO (ctx=<optimized out>, base=<optimized out>, key=<optimized out>, mis=<optimized out>) at /home/norbyte/hhvm/hhvm/hphp/runtime/vm/jit/minstr-helpers.h:144
#11 0x0000000005000b6e in ?? ()
#12 0x00007fffd7a7ffc0 in ?? ()
#13 0x0000000005000278 in ?? ()
#14 0x0000000002d5d616 in enterTCHelper () at /home/norbyte/hhvm/hhvm/hphp/runtime/vm/jit/translator-asm-helpers.S:60
#15 0x000000000191f70a in HPHP::jit::x64::(anonymous namespace)::BackEnd::enterTCHelper (this=<optimized out>, start=<optimized out>, stashedAR=<optimized out>) at /home/norbyte/hhvm/hhvm/hphp/runtime/vm/jit/back-end-x64.cpp:109
#16 0x0000000001df9a37 in HPHP::jit::MCGenerator::enterTC (this=0x7fffe8fa5900, start=0x5000270 "H\211\357\350\357l\255\374\377\340\017\v\017\v\314\314UH\211\345\377w\b\377\067H\307", <incomplete sequence \306>, stashedAR=stashedAR@entry=0x0)
    at /home/norbyte/hhvm/hhvm/hphp/runtime/vm/jit/mc-generator.cpp:1091
#17 0x0000000001761f1b in HPHP::jit::MCGenerator::enterTCAfterPrologue (this=<optimized out>, start=<optimized out>) at /home/norbyte/hhvm/hhvm/hphp/runtime/vm/jit/mc-generator.h:224
#18 0x00000000017a5c58 in HPHP::enterVMAtFunc (enterFnAr=enterFnAr@entry=0x7fffd7a7ffc0, stk=stk@entry=HPHP::Untrimmed, varEnv=varEnv@entry=0x7fffd78074f0) at /home/norbyte/hhvm/hhvm/hphp/runtime/vm/bytecode.cpp:2038
#19 0x00000000017a6438 in HPHP::enterVM (ar=ar@entry=0x7fffd7a7ffc0, stk=stk@entry=HPHP::Untrimmed, resumable=resumable@entry=0x0, exception=exception@entry=0x0, varEnv=varEnv@entry=0x7fffd78074f0) at /home/norbyte/hhvm/hhvm/hphp/runtime/vm/bytecode.cpp:2098
#20 0x0000000002d5cd06 in HPHP::ExecutionContext::invokeFunc (this=this@entry=0x7fffd7806030, retptr=retptr@entry=0x7fffd83fa500, f=<optimized out>, args_=..., this_=this_@entry=0x0, cls=cls@entry=0x0, varEnv=0x7fffd78074f0, invName=invName@entry=0x0, 
    flags=flags@entry=HPHP::ExecutionContext::InvokePseudoMain) at /home/norbyte/hhvm/hhvm/hphp/runtime/vm/bytecode.cpp:2267
#21 0x0000000002d91d7a in HPHP::ExecutionContext::invokeUnit (this=0x7fffd7806030, retval=retval@entry=0x7fffd83fa500, unit=0x7fffe7ff2680) at /home/norbyte/hhvm/hhvm/hphp/runtime/vm/bytecode.cpp:2421
#22 0x00000000014ae48a in HPHP::invoke_file_impl (res=..., path=..., once=once@entry=true, currentDir=currentDir@entry=0x2dadb63 "") at /home/norbyte/hhvm/hhvm/hphp/runtime/base/builtin-functions.cpp:707
#23 0x0000000002d91c79 in HPHP::invoke_file (s=..., once=once@entry=true, currentDir=currentDir@entry=0x2dadb63 "") at /home/norbyte/hhvm/hhvm/hphp/runtime/base/builtin-functions.cpp:720
#24 0x00000000014b05f4 in HPHP::include_impl_invoke (file=..., once=once@entry=true, currentDir=currentDir@entry=0x2dadb63 "") at /home/norbyte/hhvm/hhvm/hphp/runtime/base/builtin-functions.cpp:731
#25 0x000000000156af9a in HPHP::hphp_invoke (context=context@entry=0x7fffd7806030, cmd="*REMOVED*", func=func@entry=false, funcParams=..., funcRet=..., reqInitFunc=..., reqInitDoc="", error=@0x7fffd83fa74f: false, 
    errorMsg="Internal Server Error", once=once@entry=true, warmupOnly=warmupOnly@entry=false, richErrorMsg=richErrorMsg@entry=false) at /home/norbyte/hhvm/hhvm/hphp/runtime/base/program-functions.cpp:1981
#26 0x00000000016a97b4 in HPHP::HttpRequestHandler::executePHPRequest (this=this@entry=0x7fffe0488380, transport=transport@entry=0x7fffe8fcea18, reqURI=..., sourceRootInfo=..., cacheableDynamicContent=<optimized out>)
    at /home/norbyte/hhvm/hhvm/hphp/runtime/server/http-request-handler.cpp:476
#27 0x00000000016abad4 in HPHP::HttpRequestHandler::handleRequest (this=0x7fffe0488380, transport=0x7fffe8fcea18) at /home/norbyte/hhvm/hhvm/hphp/runtime/server/http-request-handler.cpp:387
#28 0x00000000016f1ce2 in HPHP::ServerWorker<std::shared_ptr<HPHP::FastCGIJob>, HPHP::FastCGITransportTraits>::doJobImpl (this=0x7fffe0423980, job=std::shared_ptr (count 3, weak 0) 0x7fffe0bea5c8, abort=abort@entry=false)
    at /home/norbyte/hhvm/hhvm/hphp/runtime/server/server-worker.h:106
#29 0x00000000016f202b in HPHP::ServerWorker<std::shared_ptr<HPHP::FastCGIJob>, HPHP::FastCGITransportTraits>::doJob (this=<optimized out>, job=...) at /home/norbyte/hhvm/hhvm/hphp/runtime/server/server-worker.h:56
#30 0x00000000016f0700 in HPHP::JobQueueWorker<std::shared_ptr<HPHP::FastCGIJob>, HPHP::Server*, true, false, HPHP::JobQueueDropVMStack>::start (this=0x7fffe0423980) at /home/norbyte/hhvm/hhvm/hphp/util/job-queue.h:463
#31 0x00000000016ede10 in HPHP::AsyncFunc<HPHP::ServerWorker<std::shared_ptr<HPHP::FastCGIJob>, HPHP::FastCGITransportTraits> >::run_ (obj=<optimized out>) at /home/norbyte/hhvm/hhvm/hphp/util/async-func.h:213
#32 0x00000000026da11b in HPHP::AsyncFuncImpl::threadFuncImpl (this=this@entry=0x7fffe0b50500) at /home/norbyte/hhvm/hhvm/hphp/util/async-func.cpp:131
#33 0x00000000026da469 in HPHP::AsyncFuncImpl::ThreadFunc (obj=0x7fffe0b50500) at /home/norbyte/hhvm/hhvm/hphp/util/async-func.cpp:51
#34 0x00000000015ddc40 in HPHP::start_routine_wrapper (arg=0x7fffe0445f00) at /home/norbyte/hhvm/hhvm/hphp/runtime/base/thread-hooks.cpp:95
#35 0x00007fffef31adf3 in start_thread () from /lib64/libpthread.so.0
#36 0x00007fffee8291ad in clone () from /lib64/libc.so.6
```

**Comment by fredemmott** [comment]
This appears to be fixed; do you have this with more recent builds?

**Comment by Norbyte** [comment]
Yeah, seems to be fixed in 3.8+.
